File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one.java
class Solution {
    public String addOne(String s) {
        StringBuilder ans = new StringBuilder();
        int i = s.length() - 1;
        while (i >= 0 && s.charAt(i) == '1') {
            ans.insert(0, '0');
            i--;
        }
        if (i >= 0) {
            ans.insert(0, '1');
            ans.insert(0, s.substring(0, i));
            return ans.toString();
        }
        ans.insert(0, '1');
        return ans.toString();
    }
    
    public int numSteps(String s) {
        int count = 0;
        while (!s.equals("1")) {
            if (s.charAt(s.length() - 1) == '1') {
                s = addOne(s);
            } else {
                s = s.substring(0, s.length() - 1);
            }
            count++;
        }
        return count;
    }
}

File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one.py
class Solution:
    def addOne(self, s):
        ans = ""
        i = len(s) - 1
        while (i >= 0 and s[i] == "1"):
            ans = "0" + ans
            i -= 1
        if i >= 0:
            ans = "1" + ans
            ans = s[:i] + ans
            return ans
        ans = "1" + ans
        return ans
    
    def numSteps(self, s: str) -> int:
        count = 0
        while (s != "1"):
            if s[len(s) - 1] == "1":
                s = self.addOne(s)
            else:
                s = s[:len(s) - 1]
            count += 1
        return count
                
        